core/src/main/java/org/pentaho/di/metastore/MetaStoreConst.java

```diff
@@ -66,38 +70,128 @@ public class MetaStoreConst {
-      if ( metastoreLocator == null ) {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreLocator result = metastoreLocator;
+      if ( result == null ) {
         synchronized ( this ) {
           try {
             if ( metastoreLocator == null ) {
               Collection<MetastoreLocator> metastoreLocators =
                 PluginServiceLoader.loadServices( MetastoreLocator.class );
-              metastoreLocator = metastoreLocators.stream().findFirst().orElse( null );
+              metastoreLocator = result = metastoreLocators.stream().findFirst().orElse( null );
             }
           } catch ( KettlePluginException e ) {
             logger.error( "Error getting metastore locator", e );
           }
         }
       }
-      return metastoreLocator;
+      return result;
+    }
+  };
+
+  private static final MetastoreDirectoryProvider NULL_METASTORE_DIRECTORY_PROVIDER =
+    new MetastoreDirectoryProvider() {
+      @Override
+      public IMetaStore getMetastoreForDirectory(String rootFolder) {
+        return null;
+      }
+    };
+
+  private static final Supplier<MetastoreDirectoryProvider> metastoreDirectoryProviderSupplier =
+    new Supplier<MetastoreDirectoryProvider> () {
+    private volatile MetastoreDirectoryProvider metastoreDirectoryProvider;
+
+    public MetastoreDirectoryProvider get() {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreDirectoryProvider result = metastoreDirectoryProvider;
+      if ( result == null ) {
+        synchronized ( this ) {
+          try {
+            if ( metastoreDirectoryProvider == null ) {
+              Collection<MetastoreDirectoryProvider> metastoreDirectoryProviders =
+                PluginServiceLoader.loadServices( MetastoreDirectoryProvider.class );
+              metastoreDirectoryProvider = metastoreDirectoryProviders.stream().findFirst().orElse( null );
+              if ( metastoreDirectoryProvider == null ) {
+                // cache the null so we don't constantly try to look this up.
+                metastoreDirectoryProvider = result = NULL_METASTORE_DIRECTORY_PROVIDER;
+              }
+            }
+          } catch ( KettlePluginException e ) {
+            logger.error( "Error getting metastore directory provider", e );
+          }
+        }
+      }
+      return result;
+    }
+  };
+
+  private static final MetastoreBowlProvider NULL_METASTORE_BOWL_PROVIDER =
+    new MetastoreBowlProvider() {
+      @Override
+      public IMetaStore getExplicitMetastoreForBowl(Bowl bowl) {
+        return null;
+      }
+
+      @Override
+      public IMetaStore getMetastoreForBowl(Bowl bowl) {
+        return null;
+      }
+    };
+
+  private static final Supplier<MetastoreBowlProvider> metastoreBowlProviderSupplier =
+    new Supplier<MetastoreBowlProvider> () {
+    private volatile MetastoreBowlProvider metastoreBowlProvider;
```

Comment:
does this need to be volatile?

Reply:
for the double-check locking idiom, yes.